### PR TITLE
Add WSL child plugin

### PR DESCRIPTION
**Pull request by JSCU-CNI**
This PR adds support for the discovery of Windows Subsystem for Linux (WSL) child targets on Windows target(s).

### CODE REVIEW

File: dissect/target/plugins/child/hyper-v.py [hunk 1]
```diff
@@ -17,8 +16,7 @@ class HyperVChildTargetPlugin(ChildTargetPlugin):
     PATH = "sysvol/ProgramData/Microsoft/Windows/Hyper-V/data.vmcx"
 
     def check_compatible(self):
-        if not self.target.fs.path(self.PATH).exists():
-            raise UnsupportedPluginError("No data.vmcx file found")
+        return self.target.fs.path(self.PATH).exists()
```
Comment (Schamper) [inline dissect/target/plugins/child/hyper-v.py:19]: Why this change?

File: dissect/target/loader.py [hunk 2]
```diff
@@ -147,7 +147,7 @@ def find_loader(item: Path, parsed_path: Optional[urllib.parse.ParseResult] = No
         try:
             if loader.detect(item):
                 return loader
-        except ImportError as exception:
+        except (ImportError, NotImplementedError) as exception:
```
Comment (Schamper) [inline dissect/target/loader.py:150]: Can you also elaborate on this addition? Not that I'm against it, just wondering why it was added.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/loader.py (hunk 2 above)
Comment (JSCU-CNI) [reply dissect/target/loader.py:150]: Before defaulting to the default `RawLoader`, `VBLoader` tries to load the `vhdx` file first and breaks when trying to parse it.

File: dissect/target/plugins/child/hyper-v.py (hunk 1 above)
Comment (JSCU-CNI) [reply dissect/target/plugins/child/hyper-v.py:19]: Consistency across all `child` plugins and because `_load_child_plugins()` in `target.py` checks boolean values instead of catching any `UnsupportedPluginError` exceptions.

File: dissect/target/loader.py (hunk 2 above)
Comment (Schamper) [reply dissect/target/loader.py:150]: I'm not getting the same error when I remove this. Can you be more specific about where the error is originating from?

File: dissect/target/plugins/child/hyper-v.py (hunk 1 above)
Comment (Schamper) [reply dissect/target/plugins/child/hyper-v.py:19]: `UnsupportedPluginError` is a subclass of `PluginError` which is caught. In this case, raising an error might give a little bit more information on why it's not compatible.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/loader.py (hunk 2 above)
Comment (JSCU-CNI) [reply dissect/target/loader.py:150]: It seems like it's a race condition, have you tried running it multiple times?

Comment (Schamper) [reply dissect/target/loader.py:150]: Yes, haven't seen it yet. Can you post a stacktrace?

Comment (JSCU-CNI) [reply dissect/target/loader.py:150]:
```
$ pytest tests/test_plugins_child_wsl.py
============================================= test session starts =============================================
platform linux -- Python 3.9.5, pytest-7.2.0, pluggy-1.0.0
rootdir: /tmp/dissect.target
collected 1 item                                                                                              

tests/test_plugins_child_wsl.py F                                                                       [100%]

================================================== FAILURES ===================================================
___________________________________________ test_plugins_child_wsl ____________________________________________

self = TargetPath('C:/Users/John/AppData/Local/Packages/CanonicalGroupLimited.Ubuntu22.04LTS_79rhkp1fndgsc/LocalState/ext4.vhdx/MFT_C.bin')

    def get(self):
        try:
>           return self._entry
E           AttributeError: _entry

dissect/target/helpers/fsutil.py:660: AttributeError

During handling of the above exception, another exception occurred:

target_win_users = <Target /tmp/MockTarget-e3m3u0cj>, fs_win = <VirtualFilesystem>

    def test_plugins_child_wsl(target_win_users, fs_win):
    
        vhdx_file = absolute_path("data/ext4.vhdx")
        fs_win.map_file(
            "users/john/appdata/local/Packages/CanonicalGroupLimited.Ubuntu22.04LTS_79rhkp1fndgsc/LocalState/ext4.vhdx",
            vhdx_file,
        )
    
        target_win_users.add_plugin(WSLTargetPlugin)
    
        children = list(target_win_users.list_children())
        assert len(children) == 1
        assert children[0].type == "wsl"
        assert (
            children[0].path
            == "C:/Users/John/AppData/Local/Packages/CanonicalGroupLimited.Ubuntu22.04LTS_79rhkp1fndgsc/LocalState/ext4.vhdx"  # noqa E501
        )
    
>       child_targets = list(target_win_users.open_children())

tests/test_plugins_child_wsl.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dissect/target/target.py:383: in open_children
    target = self.open_child(child.path)
dissect/target/target.py:368: in open_child
    return Target.open(self.fs.path(child))
dissect/target/target.py:222: in open
    loader_cls = loader.find_loader(path, parsed_path=parsed_path)
dissect/target/loader.py:148: in find_loader
    if loader.detect(item):
dissect/target/loaders/vb.py:10: in detect
    mft_exists = path.joinpath("MFT_C.bin").exists()
dissect/target/helpers/fsutil.py:761: in exists
    self.get().stat()
dissect/target/helpers/fsutil.py:662: in get
    self._entry = self._fs.get(str(self))  # noqa
dissect/target/filesystem.py:1183: in get
    entries.append(self._get_from_entry(path, sub_entry))
dissect/target/filesystem.py:1203: in _get_from_entry
    entry = entry.get(part)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <MappedFile 'users/john/appdata/local/Packages/CanonicalGroupLimited.Ubuntu22.04LTS_79rhkp1fndgsc/LocalState/ext4.vhdx'>
path = 'MFT_C.bin'

    def get(self, path: str) -> FilesystemEntry:
        """Retrieve a FilesystemEntry relative to this entry.
    
        Args:
            path: The path relative to this filesystem entry.
    
        Returns:
            A relative FilesystemEntry.
        """
>       raise NotImplementedError()
E       NotImplementedError

dissect/target/filesystem.py:424: NotImplementedError
---------------------------------------------- Captured log call ----------------------------------------------
WARNING  dissect.target.loader:loader.py:151 Failed to import <lazyattr dissect.target.loaders.asdf.AsdfLoader loaded=False>
============================================== warnings summary ===============================================
tests/test_plugins_child_wsl.py::test_plugins_child_wsl
tests/test_plugins_child_wsl.py::test_plugins_child_wsl
  /tmp/.local/lib/python3.9/site-packages/flow/record/fieldtypes/__init__.py:471: DeprecationWarning: Do not use class uri(...) for filesystem paths, use class path(...)
    warnings.warn(

tests/test_plugins_child_wsl.py::test_plugins_child_wsl
tests/test_plugins_child_wsl.py::test_plugins_child_wsl
  /tmp/.local/lib/python3.9/site-packages/flow/record/fieldtypes/__init__.py:462: DeprecationWarning: Do not use class uri(...) for filesystem paths, use class path(...)
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================== short test summary info ===========================================
FAILED tests/test_plugins_child_wsl.py::test_plugins_child_wsl - NotImplementedError
======================================== 1 failed, 4 warnings in 0.65s ========================================
```

Comment (Schamper) [reply dissect/target/loader.py:150]: That looks more like a bug in `VirtualFilesystem`. Also can you add that test to the PR ;)

Comment (Schamper) [reply dissect/target/loader.py:150]: Should be fixed with #155

File: dissect/target/plugins/child/hyper-v.py (hunk 1 above)
Comment (Schamper) [reply dissect/target/plugins/child/hyper-v.py:19]: It already catches those exceptions though:
https://github.com/fox-it/dissect.target/blob/main/dissect/target/target.py#L345-L350

Although I agree the messaging could be improved (i.e. a try/except being moved closer to the `check_compatible` with a more descriptive message), the functionality is already the same.

Comment (JSCU-CNI) [reply dissect/target/plugins/child/hyper-v.py:19]: Thanks for explaining. Although the exception message will be 'broken child plugin' when in reality it's just a non-compatible child plugin.

Comment (Schamper) [reply dissect/target/plugins/child/hyper-v.py]: Please check https://github.com/fox-it/dissect.target/pull/148/commits/806d4c41b0890cc4c4af54d1955206f0106bd917. I've changed the logging around a bit for child plugins so it should be a lot cleaner and clear about what's happening.

File: dissect/target/loader.py (hunk 2 above)
Comment (JSCU-CNI) [reply dissect/target/loader.py]: That's great. Removed the `NotImplementedError` in [34541c8](https://github.com/fox-it/dissect.target/pull/148/commits/34541c861b00450d146bdf9f6721a78f53f3f101). Added the test in [1ce5f50](https://github.com/fox-it/dissect.target/pull/148/commits/1ce5f508f09b5acc8226448a7b1d1dba28a098de), but we can currently not provide a `ext4.vhdx` example file. Could you create a dummy `ext4.vhdx` file and upload it using git lfs?

Comment (Schamper) [reply dissect/target/loader.py]: Thanks, I've changed the test to only test the functionality of the WSL plugin itself, so no full `ext4.vhdx` necessary.

**Review by Schamper — APPROVED** [review]
